src/transpile/cacheblocking.hpp

```diff
@@ -0,0 +1,684 @@
+  int nq,nb;
+  bool exist;
+  uint_t pos_begin,num_gates_added;
+
+  pos_begin = out.size();
+  num_gates_added = 0;
+
+  i = 0;
+  //add chunk swap and block ops (if blocking is enabled)
+  if(blocking_enabled_){
+    while(i <ops.size()){
+      if(ops[i].type == Operations::OpType::sim_op){
+        out.push_back(ops[i]);
+      }
+      else if(ops[i].type == Operations::OpType::gate && ops[i].name == "swap_chunk"){
+        out.push_back(ops[i]);
+      }
+      else{
+        break;
+      }
+    }
+  }
+
+  //gather blocked gates
+  Operations::Op blk_begin;
+  blk_begin.type = Operations::OpType::sim_op;
+  blk_begin.name = "begin_register_blocking";
+  out.push_back(blk_begin);
+
+  while(i < ops.size()){
+    if(ops[i].type == Operations::OpType::gate || ops[i].type == Operations::OpType::matrix){
+      if((ops[i].qubits.size() > 1 && ops[i].type == Operations::OpType::matrix) || ops[i].name == "pauli"){
+        queue.push_back(ops[i]);
+      }
+      else{
+        //check if qubits of operation is included in the queued operations
+        exist = false;
+        for(j=0;j<queue.size();j++){
+          for(iq=0;iq<ops[i].qubits.size();iq++){
+            for(jq=0;jq<queue[j].qubits.size();jq++){
+              if(ops[i].qubits[iq] == queue[j].qubits[jq]){
+                exist = true;
+                break;
+              }
+            }
+          }
+          if(exist)
+            break;
+        }
+        if(exist){
+          queue.push_back(ops[i]);
+        }
+        else{
+          if(is_diagonal_op(ops[i])){
+            //diagonal gate can be applied
+            out.push_back(ops[i]);
+            num_gates_added++;
+          }
+          else{
+            exist = false;
+            iq = ops[i].qubits[ops[i].qubits.size()-1]; //block target bit
+            nq = blockedQubits.size();
+            for(j=0;j<nq;j++){
+              if(iq == blockedQubits[j]){
+                exist = true;
+                break;
+              }
+            }
+            if(exist){
+              out.push_back(ops[i]);
+              num_gates_added++;
+            }
+            else{
+              if(nq == gpu_blocking_bits_){
+                queue.push_back(ops[i]);
+              }
+              else{
+                blockedQubits.push_back(iq);
+                out.push_back(ops[i]);
+                num_gates_added++;
+              }
+            }
+          }
+        }
+      }
+    }
+    else{
+      queue.push_back(ops[i]);
+    }
+    i++;
+  }
+
+  if(out.size() > pos_begin + 1){
+    out[pos_begin].qubits = blockedQubits;
+
+    Operations::Op blk_end;
+    blk_end.type = Operations::OpType::sim_op;
+    blk_end.name = "end_register_blocking";
+    out.push_back(blk_end);
+  }
+  else{
+    out.pop_back();
+  }
+
+  return num_gates_added;
+}
+
+bool CacheBlocking::isCrossQubitsOp(Operations::Op& op, reg_t& qubits) const
+{
+  bool is_cross_qubits = false;
+  bool is_pauli_str = false;
```

Comment:
In this function and in `is_diagonal_op` function you check if it `is_pauli_str`. It'd be better to have that check isolated in a function and call it from both `isCrossQubitsOp` and `is_diagonal_op`.